File: stacks/typescript/src/middleware/rate-limiter.ts
import { Request, Response, NextFunction } from 'express';

interface RateLimiterOptions {
  windowMs: number;
  maxRequests: number;
}

interface ClientRecord {
  count: number;
  resetTime: number;
}

export function rateLimiter(options: RateLimiterOptions) {
  const clients = new Map<string, ClientRecord>();

  // Cleanup expired entries periodically
  const cleanupInterval = setInterval(() => {
    const now = Date.now();
    for (const [key, record] of clients) {
      if (now > record.resetTime) {
        clients.delete(key);
      }
    }
  }, options.windowMs);

  // Allow cleanup timer to not prevent process exit
  if (cleanupInterval.unref) {
    cleanupInterval.unref();
  }

  return (req: Request, res: Response, next: NextFunction): void => {
    const clientIp = req.ip ?? req.socket.remoteAddress ?? 'unknown';
    const now = Date.now();
    const record = clients.get(clientIp);

    if (!record || now > record.resetTime) {
      clients.set(clientIp, { count: 1, resetTime: now + options.windowMs });
      next();
      return;
    }

    if (record.count >= options.maxRequests) {
      res.status(429).json({
        error: 'Too many requests',
        retryAfter: Math.ceil((record.resetTime - now) / 1000),
      });
      return;
    }

    record.count++;
    next();
  };
}

File: stacks/python/app/middleware/rate_limiter.py
import time
from collections import defaultdict
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import JSONResponse


class RateLimiterMiddleware(BaseHTTPMiddleware):
    def __init__(self, app, max_requests: int = 100, window_seconds: int = 60):
        super().__init__(app)
        self.max_requests = max_requests
        self.window_seconds = window_seconds
        self.clients: dict[str, list[float]] = defaultdict(list)

    async def dispatch(self, request: Request, call_next):
        client_ip = request.client.host if request.client else "unknown"
        now = time.monotonic()

        # Clean old entries
        self.clients[client_ip] = [
            t for t in self.clients[client_ip]
            if now - t < self.window_seconds
        ]

        if len(self.clients[client_ip]) >= self.max_requests:
            oldest = self.clients[client_ip][0]
            retry_after = int(self.window_seconds - (now - oldest)) + 1
            return JSONResponse(
                status_code=429,
                content={"error": "Too many requests", "retry_after": retry_after},
            )

        self.clients[client_ip].append(now)
        return await call_next(request)
